PFW-1362 Implement SpoolJoin + FINDA filament runout

gudnimg opened the pull request:
The pull request rewrites SpoolJoin implementation (no drastic changes to old behaviour though) and now the firmware uses the FINDA sensor for filament runout detection when the MMU2S is used. Works pretty well in my testing so far.

* Commit ca67fa9cbedd6d2cfa80eb014d9f1d6f7e77f16c fixes an important issue where FINDA values was not being written to findaPressed variable.

gudnimg committed:
PFW-1362 Initial commit

gudnimg committed:
Rename files to SpoolJoin

gudnimg committed:
Remove unused CMake tests

gudnimg committed:
PFW-1362 Fsensor status is not required for SpoolJoin

gudnimg committed:
PFW-1362 Rename settings for consistency

gudnimg committed:
PFW-1362 Fsensor is not required for Spooljoin

Since the Fsensor is not required to use SpoolJoin,
it makes little sense to me that we need the fsensor
to "enable" SpoolJoin

gudnimg committed:
PFW-1362 Remove "NA" EEPROM status

gudnimg committed:
PFW-1362 Show if SpoolJoin is enabled or disabled at bootup

gudnimg committed:
PFW-1362 Add a log when FINDA filament runout is triggered

gudnimg committed:
PFW-1362 temporary debug messages

gudnimg committed:
PFW-1362 temporarily don't use SpoolJoin in Tcodes

I think in the current implementation, this will cause more harm than good.
But we can improve SpoolJoin later

gudnimg committed:
Fix an issue where findaPressed is never set

This caused instant FINDA runout when the print starts. Now fixed :)

gudnimg committed:
PFW-1362 SpoolJoin needs to known what the previous slot was

DRracer commented on Firmware/SpoolJoin.h:
```diff
@@ -0,0 +1,59 @@
+/// @file
```
nice, clean code and well documented :+1:

DRracer commented on Firmware/SpoolJoin.cpp:
```diff
@@ -0,0 +1,80 @@
+    SERIAL_ECHOPGM("SpoolJoin is ");
+    if (isSpoolJoinEnabled())
+    {
+        SERIAL_ECHOLNPGM("enabled");
```
:thinking: don't we have some already existing message (messages.h) which could be used instead of these "enabled/disabled" texts?

gudnimg replied:
We have "Off" and "On" those can serve the same purpose.

gudnimg replied:
Done. Saves 16 bytes. I used `_O()` on the strings since we don't care about translations in this case.

DRracer commented on Firmware/SpoolJoin.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void SpoolJoin::initSpoolJoinStatus()
+{
+    EEPROM currentStatus = (EEPROM)eeprom_read_byte((uint8_t*)EEPROM_AUTO_DEPLETE);
```
another topic for potential optimization - many variables read from the eeprom are checked for unitialized values. We could try implementing a non-inline function:
`uint8_t eeprom_get_byte(uint8_t * address, uint8_t defaultValue)`

We could save some space doing the check for defaults at one spot and having always a valid value as a result. This function could be updating the eeprom cell with the default as well if necessary.

DRracer commented on Firmware/SpoolJoin.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void SpoolJoin::toggleSpoolJoin()
+{
+    if (eeprom_read_byte((uint8_t*)EEPROM_AUTO_DEPLETE) == (uint8_t)EEPROM::Disabled)
```
potential optimization - eeprom_toggle()? Do we have similar implementation in the code base multiple times?

gudnimg replied:
I'm not sure we do. Probably a good idea though.

DRracer commented on Firmware/SpoolJoin.cpp:
```diff
@@ -0,0 +1,80 @@
+    }
+}
+
+uint8_t SpoolJoin::isSpoolJoinEnabled()
```
Can't it be `bool`?

gudnimg replied:
Done. You're right, I'm just used to using `uint8_t` for everything in C

DRracer commented on Firmware/SpoolJoin.cpp:
```diff
@@ -0,0 +1,80 @@
+
+uint8_t SpoolJoin::nextSlot()
+{
+    SERIAL_ECHOPGM("SpoolJoin: Changing slot from ");
```
let's optimize this message :wink: 
`SpoolJoin: x -> y`

gudnimg replied:
Done, saved 20 bytes!

gudnimg committed:
PFW-1362 Use PROGMEM messages

The main purpose of these two messages is just to see if SpoolJoin is enabled or not.
"On" and "Off" are descriptive enough.

Saves 16 bytes of flash

DRracer commented on Firmware/Tcodes.cpp:
```diff
@@ -57,14 +57,14 @@ void TCodes(char *const strchr_pointer, uint8_t codeValue) {
+        if (strchr_pointer[index] == '?') {
+            selectedSlot = TCodeChooseFromMenu();
+        /*} else if (MMU2::mmu2.Enabled() && SpoolJoin::spooljoin.isSpoolJoinEnabled()) {
+            // TODO: What if the next slot has no filament?
```
I think the original idea was to select next slot again of there was no filament, but we can skip that for now.

gudnimg committed:
PFW-1362 Optimise debug message

Saves 20 bytes of flash :)

DRracer commented on Firmware/mmu2_protocol_logic.cpp:
```diff
@@ -316,6 +316,7 @@ StepStatus ProtocolLogic::CommandStep() {
         SendFINDAQuery();
         return Processing;
     case ScopeState::FINDAReqSent:
+        findaPressed = rsp.paramValue;
```
:man_facepalming: I forgot about this one ... moreover - there are 2 idential pieces of code (in CommandStep and in IdleStep). May be we could make them one function if it saves some bytes.

gudnimg committed:
PFW-1362 Change isSpoolJoinEnabled to return bool

DRracer commented on Firmware/mmu2_protocol_logic.cpp:
```diff
@@ -316,6 +316,7 @@ StepStatus ProtocolLogic::CommandStep() {
     case ScopeState::FINDAReqSent:
+        findaPressed = rsp.paramValue;
         SendReadRegister(4, ScopeState::StatisticsSent);
         scopeState = ScopeState::StatisticsSent;
```
even this line can be removed, I think SendReadRegister does that already

gudnimg replied:
Done, this saved 2 bytes of flash. See 279aa8e

DRracer reviewed (approved):
well done :+1: , just think about the optimization ideas, may be we can shave off some bytes.

DRracer commented on Firmware/SpoolJoin.h:
```diff
@@ -31,6 +31,7 @@ class SpoolJoin {
     static void toggleSpoolJoin();
 
     /// @brief Check if SpoolJoin is enabled
+    /// @returns 1 if enabled, 0 if disabled
```
probably true/false now (not 1/0)

gudnimg replied:
Done

gudnimg commented:
Rebased the PR with respect to `MK3_MMU2` branch

gudnimg commented:
If the MMU is disabled. The firmware will still trigger a FINDA runout (even though there is no FINDA sensor). I was kind of expecting this :)

EDIT: Fixed here [1524aae](https://github.com/prusa3d/Prusa-Firmware/pull/3586/commits/1524aae2bffa69f1e6e14ceb9d4bf6a890e7dccc)

gudnimg committed:
PFW-1362 Make sure FINDA runout is impossible with MMU disabled

Since mmu_loop() is called in the main loop() function
when the MMU is disabled, we need to utilise the
MMU protocol layer to prevent FINDA runout from happening
if MMU is disabled.

We also need to keep in mind we probably can't trigger
a FINDA runout in the middle of a Q0 query.
So my solution now is to wait for the "Finished" state
and check if the FINDA is off during printing.

When "Finished" state appears, the FINDA value should be up
to date.

gudnimg committed:
Optimisation: scopeState is already set by SendReadRegister

Saves 2 bytes of flash

DRracer commented on Firmware/mmu2.cpp:
```diff
@@ -731,6 +734,9 @@ StepStatus MMU2::LogicStep() {
     StepStatus ss = logic.Step();
     switch (ss) {
     case Finished:
+        // At this point it is safe to trigger a runout and not interrupt the MMU protocol
```
this is actually a good point :+1:
